### ISSUE TITLE
opt: add support for LATERAL in FROM clause

### ISSUE THREAD

**Issue by jseldess** [cockroachdb/docs#4882]
PR: https://github.com/cockroachdb/cockroach/pull/36613

From release notes:
> The `LATERAL` keyword in a `FROM` clause is now supported. [#36613][#36613] {% comment %}doc{% endcomment %}

**Comment by rmloveland** [comment]
@justinj I have questions

If I set up this data set

```
CREATE DATABASE IF NOT EXISTS employees;
USE employees;
IMPORT MYSQLDUMP 'https://s3-us-west-1.amazonaws.com/cockroachdb-movr/datasets/employees-db/mysqldump/employees-full.sql.gz';
```

and then try to adapt the `LATERAL` example from [lateral derived tables](https://www.ibm.com/support/knowledgecenter/en/SSGU8G_11.70.0/com.ibm.sqls.doc/ids_sqs_0254.htm) as follows, it doesn't work.

`SELECT d.dept_no, empinfo.avgsal, empinfo.empcount FROM departments d, LATERAL (SELECT AVG(salary) FROM employees,salaries WHERE employees.emp_no = salaries.emp_no AS avgsal, COUNT(*) FROM departments,dept_emp WHERE dept_emp.dept_no=departments.dept_no) AS empinfo;`

The above breaks with 

```
-- invalid syntax: statement ignored: at or near "as": syntax error
-- DETAIL: source SQL:
-- SELECT d.dept_no, empinfo.avgsal, empinfo.empcount FROM departments d, LATERAL (SELECT AVG(salary) FROM employees,salaries WHERE employees.emp_no = salaries.emp_no AS avgsal, COUNT(*) FROM departments,dept_emp WHERE dept_emp.dept_no=departments.dept_no) AS empinfo
--                                                                                                                                                                     ^
```

However if I rewrite the query to use a leading `WITH` so it at least parses according to `cockroach sqlfmt`, it fails like so:

```sql
WITH
	empinfo
		AS (
			SELECT
				count(*)
			FROM
				departments, dept_emp
			WHERE
				dept_emp.dept_no = departments.dept_no
		)
SELECT
	d.dept_no, empinfo.avgsal, empinfo.empcount
FROM
	departments AS d,
	LATERAL (
		SELECT
			avg(salary)
		FROM
			employees, salaries
		WHERE
			employees.emp_no = salaries.emp_no
	);
```

output:

```
pq: no data source matches prefix: empinfo
```

However, it appears that I did just create that data source in the `WITH empinfo ...` clause?

Is this a user error on my part or nuance in what bits of "`LATERAL` in `FROM` clause" we do / don't support?

**Comment by justinj** [comment]
The first is failing because your `AS` clause is misplaced—you can't alias a `WHERE` clause, I think you have a paren in the wrong place somewhere. The second is because the `WITH` introduces a new data source, but you never select from it, so you need to include it in your `FROM` clause somewhere. I'm not sure that will work though, I think if you fix the location of your `AS` in the first one that should work.

**Comment by rmloveland** [comment]
Thanks Justin!  I got something ~working from the above.

tl;dr: Is the query below failing because I'm using the default memory setting (128MiB)?

Next question.  I tried to run the following query on the same data set (this is the query from the [Correlated subqueries "limitations" section](https://www.cockroachlabs.com/docs/dev/subqueries.html#limitations)). 

```sql
SELECT
 e.last_name, s.salary, noise
FROM
 employees AS e,
 salaries AS s,
 generate_series(0, s.salary, 10000) AS noise
WHERE
 e.emp_no = s.emp_no
ORDER BY
 s.salary DESC
LIMIT
 10;
```

 It does not fail with an error but seems to hang a single node local cluster / peg a core at 100% for ~45 minutes' CPU time.  Has not returned any response to the CLI.

![Screenshot 2019-11-04 16 45 04](https://user-images.githubusercontent.com/1312291/68160767-7f3a2a00-ff22-11e9-9ef0-4af860a0b8e7.png)

Should I be using a bigger cluster with `roachprod` or something?  

Here are some screenshots from the Admin UI which seem weird to my inexpert eyes

e.g. "no queries at all, oh except except 1 distributed query" O RLY

![Screenshot 2019-11-04 16 47 49](https://user-images.githubusercontent.com/1312291/68161006-f079dd00-ff22-11e9-8e3b-6dc3c7c09f54.png)

Furthermore, the query does not appear at all on the statements page.

It's not a very big db, about 77MiB FWIW (but that might not matter, IDK). 

![Screenshot 2019-11-04 16 49 41](https://user-images.githubusercontent.com/1312291/68161139-346ce200-ff23-11e9-99f9-f95fb47ba47b.png)

OH but I'm running with the default memory (128MiB right?).  OK I'll try "downloading more RAM" and report back.  Still I suppose one could argue that some of the data set should fit in RAM?

Sincerely,
![sciencedog](https://user-images.githubusercontent.com/1312291/68161582-253a6400-ff24-11e9-9e54-d3bb736b37ca.jpg)
